<a href="https://colab.research.google.com/github/gong-aipel/AIFFEL_quest-cr/blob/main/python/6%EC%9E%A5_%ED%94%84%EB%A1%9C%EC%A0%9D%ED%8A%B8123.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import sklearn

print(sklearn.__version__)

1.6.0


In [9]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

from sklearn.metrics import accuracy_score

In [3]:
import pandas as pd



In [4]:
# (1) load_breast_cancer 메서드로 데이터 로드
data = load_breast_cancer()

# (2) Feature Data 지정하기
features = pd.DataFrame(data.data, columns=data.feature_names)

# (3) Label Data 지정하기
labels = pd.Series(data.target, name="label")

# (4) Target Names 출력해 보기
print("Target Names:", data.target_names)

# (5) 데이터 Describe 해 보기
print("\nFeature Data Description:")
print(features.describe())

# (6) train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(
    features, labels, test_size=0.2, random_state=42
)

print("\nTrain and Test split completed:")
print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

Target Names: ['malignant' 'benign']

Feature Data Description:
       mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630    

In [5]:
from sklearn.model_selection import train_test_split

# train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(
    features,  # Feature Data
    labels,    # Label Data
    test_size=0.2,  # 테스트 데이터 비율 (20%)
    random_state=42  # 결과 재현을 위한 랜덤 시드
)

print("Train and Test data shapes:")
print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

Train and Test data shapes:
X_train shape: (455, 30)
X_test shape: (114, 30)
y_train shape: (455,)
y_test shape: (114,)


In [11]:
# 데이터 로드
data = load_breast_cancer()
X = data.data
y = data.target

# train-test 데이터 나누기
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 유방암 데이터 확인
print(f"Feature data shape: {X_train.shape}")
print(f"Label data shape: {y_train.shape}")

Feature data shape: (455, 30)
Label data shape: (455,)


In [12]:
from sklearn.tree import DecisionTreeClassifier

# 모델 생성
decision_tree = DecisionTreeClassifier(random_state=42)

# 학습
decision_tree.fit(X_train, y_train)

# 예측
y_pred = decision_tree.predict(X_test)

# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print(f"Decision Tree Accuracy: {accuracy:.2f}")

Decision Tree Accuracy: 0.95


In [14]:
from sklearn.ensemble import RandomForestClassifier

# 모델 생성
random_forest = RandomForestClassifier(random_state=42)

# 학습
random_forest.fit(X_train, y_train)

# 예측
y_pred = random_forest.predict(X_test)

# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print(f"Random Forest Accuracy: {accuracy:.2f}")

Random Forest Accuracy: 0.96


In [15]:
from sklearn.svm import SVC

# 모델 생성
svm = SVC(random_state=42)

# 학습
svm.fit(X_train, y_train)

# 예측
y_pred = svm.predict(X_test)

# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print(f"SVM Accuracy: {accuracy:.2f}")

SVM Accuracy: 0.95


In [16]:
from sklearn.linear_model import SGDClassifier

# 모델 생성
sgd = SGDClassifier(random_state=42)

# 학습
sgd.fit(X_train, y_train)

# 예측
y_pred = sgd.predict(X_test)

# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print(f"SGD Classifier Accuracy: {accuracy:.2f}")

SGD Classifier Accuracy: 0.90


In [17]:
from sklearn.linear_model import LogisticRegression

# 모델 생성
logistic_regression = LogisticRegression(max_iter=1000, random_state=42)

# 학습
logistic_regression.fit(X_train, y_train)

# 예측
y_pred = logistic_regression.predict(X_test)

# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print(f"Logistic Regression Accuracy: {accuracy:.2f}")

Logistic Regression Accuracy: 0.96


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [20]:
from sklearn.metrics import classification_report, f1_score

# 모델 리스트와 이름
models = {
    "Decision Tree": DecisionTreeClassifier(random_state=42),
    "Random Forest": RandomForestClassifier(random_state=42),
    "SVM": SVC(random_state=42),
    "SGD Classifier": SGDClassifier(random_state=42),
    "Logistic Regression": LogisticRegression(max_iter=1000, random_state=42)}

    # 각 모델의 결과 저장
results = {}

# 모든 모델 평가
for model_name, model in models.items():
    # 학습
    model.fit(X_train, y_train)
    # 예측
    y_pred = model.predict(X_test)
    # 평가 결과 저장
    f1 = f1_score(y_test, y_pred)
    results[model_name] = {
        "f1_score": f1,
        "classification_report": classification_report(y_test, y_pred, target_names=data.target_names)
    }

# 가장 높은 F1 점수를 가진 모델 선택
best_model = max(results, key=lambda x: results[x]["f1_score"])

# 결과 출력
print(f"Best Model: {best_model}")
print(f"F1 Score: {results[best_model]['f1_score']:.2f}")
print("\nClassification Report:\n")
print(results[best_model]["classification_report"])

Best Model: Random Forest
F1 Score: 0.97

Classification Report:

              precision    recall  f1-score   support

   malignant       0.98      0.93      0.95        43
      benign       0.96      0.99      0.97        71

    accuracy                           0.96       114
   macro avg       0.97      0.96      0.96       114
weighted avg       0.97      0.96      0.96       114



/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


대부분의 데이터에서 설명력이 좋게 나오는 결과와 동일하게 붓꽃 데이터도 랜덤 포레스트가 97%로 제일 좋게 나왔다.

In [21]:
import sklearn

print(sklearn.__version__)

1.6.0


In [22]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [23]:
data = load_digits()

In [24]:
# (2) Feature Data 지정하기
features = pd.DataFrame(data.data, columns=data.feature_names)

# (3) Label Data 지정하기
labels = pd.Series(data.target, name="label")

# (4) Target Names 출력해 보기
print("Target Names:", data.target_names)

# (5) 데이터 Describe 해 보기
print("\nFeature Data Description:")
print(features.describe())

# (6) train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(
    features, labels, test_size=0.2, random_state=42
)

print("\nTrain and Test split completed:")
print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

Target Names: [0 1 2 3 4 5 6 7 8 9]

Feature Data Description:
       pixel_0_0    pixel_0_1    pixel_0_2    pixel_0_3    pixel_0_4  \
count     1797.0  1797.000000  1797.000000  1797.000000  1797.000000   
mean         0.0     0.303840     5.204786    11.835838    11.848080   
std          0.0     0.907192     4.754826     4.248842     4.287388   
min          0.0     0.000000     0.000000     0.000000     0.000000   
25%          0.0     0.000000     1.000000    10.000000    10.000000   
50%          0.0     0.000000     4.000000    13.000000    13.000000   
75%          0.0     0.000000     9.000000    15.000000    15.000000   
max          0.0     8.000000    16.000000    16.000000    16.000000   

         pixel_0_5    pixel_0_6    pixel_0_7    pixel_1_0    pixel_1_1  ...  \
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000  ...   
mean      5.781859     1.362270     0.129661     0.005565     1.993879  ...   
std       5.666418     3.325775     1.037383     0.

In [29]:
data = load_digits()

In [33]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, f1_score

# 데이터 로드
digits = load_digits()
X = digits.data  # Feature 데이터
y = digits.target  # Label 데이터

# 학습 데이터와 테스트 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"Feature data shape: {X_train.shape}")
print(f"Label data shape: {y_train.shape}")

Feature data shape: (1437, 64)
Label data shape: (1437,)


In [34]:
from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state=42)
decision_tree.fit(X_train, y_train)
y_pred_dt = decision_tree.predict(X_test)

accuracy_dt = accuracy_score(y_test, y_pred_dt)
f1_dt = f1_score(y_test, y_pred_dt, average='macro')
print(f"Decision Tree - Accuracy: {accuracy_dt:.2f}, F1 Score: {f1_dt:.2f}")

Decision Tree - Accuracy: 0.84, F1 Score: 0.84


In [35]:
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=42)
random_forest.fit(X_train, y_train)
y_pred_rf = random_forest.predict(X_test)

accuracy_rf = accuracy_score(y_test, y_pred_rf)
f1_rf = f1_score(y_test, y_pred_rf, average='macro')
print(f"Random Forest - Accuracy: {accuracy_rf:.2f}, F1 Score: {f1_rf:.2f}")

Random Forest - Accuracy: 0.97, F1 Score: 0.97


In [36]:
from sklearn.svm import SVC

svm = SVC(random_state=42)
svm.fit(X_train, y_train)
y_pred_svm = svm.predict(X_test)

accuracy_svm = accuracy_score(y_test, y_pred_svm)
f1_svm = f1_score(y_test, y_pred_svm, average='macro')
print(f"SVM - Accuracy: {accuracy_svm:.2f}, F1 Score: {f1_svm:.2f}")

SVM - Accuracy: 0.99, F1 Score: 0.99


In [37]:
from sklearn.linear_model import SGDClassifier

sgd = SGDClassifier(random_state=42)
sgd.fit(X_train, y_train)
y_pred_sgd = sgd.predict(X_test)

accuracy_sgd = accuracy_score(y_test, y_pred_sgd)
f1_sgd = f1_score(y_test, y_pred_sgd, average='macro')
print(f"SGD Classifier - Accuracy: {accuracy_sgd:.2f}, F1 Score: {f1_sgd:.2f}")

SGD Classifier - Accuracy: 0.96, F1 Score: 0.96


In [38]:
from sklearn.linear_model import LogisticRegression

logistic_regression = LogisticRegression(max_iter=1000, random_state=42)
logistic_regression.fit(X_train, y_train)
y_pred_lr = logistic_regression.predict(X_test)

accuracy_lr = accuracy_score(y_test, y_pred_lr)
f1_lr = f1_score(y_test, y_pred_lr, average='macro')
print(f"Logistic Regression - Accuracy: {accuracy_lr:.2f}, F1 Score: {f1_lr:.2f}")

Logistic Regression - Accuracy: 0.97, F1 Score: 0.98


In [39]:
# 결과 저장
results = {
    "Decision Tree": f1_dt,
    "Random Forest": f1_rf,
    "SVM": f1_svm,
    "SGD Classifier": f1_sgd,
    "Logistic Regression": f1_lr
}

best_model = max(results, key=results.get)
print("\nBest Model:")
print(f"{best_model} with F1 Score: {results[best_model]:.2f}")


Best Model:
SVM with F1 Score: 0.99


손글씨 데이터는 예외적으로 SVM 분석이 0.99로 랜덤포레스트인 0.97보다 높아서 이 경우는 SVM으로 선택하는 것이 바람직하다고 할 수 있겠다.

In [40]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [41]:
data = load_wine()

In [42]:
# (2) Feature Data 지정하기
features = pd.DataFrame(data.data, columns=data.feature_names)

# (3) Label Data 지정하기
labels = pd.Series(data.target, name="label")

# (4) Target Names 출력해 보기
print("Target Names:", data.target_names)

# (5) 데이터 Describe 해 보기
print("\nFeature Data Description:")
print(features.describe())

# (6) train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(
    features, labels, test_size=0.2, random_state=42
)

print("\nTrain and Test split completed:")
print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

Target Names: ['class_0' 'class_1' 'class_2']

Feature Data Description:
          alcohol  malic_acid         ash  alcalinity_of_ash   magnesium  \
count  178.000000  178.000000  178.000000         178.000000  178.000000   
mean    13.000618    2.336348    2.366517          19.494944   99.741573   
std      0.811827    1.117146    0.274344           3.339564   14.282484   
min     11.030000    0.740000    1.360000          10.600000   70.000000   
25%     12.362500    1.602500    2.210000          17.200000   88.000000   
50%     13.050000    1.865000    2.360000          19.500000   98.000000   
75%     13.677500    3.082500    2.557500          21.500000  107.000000   
max     14.830000    5.800000    3.230000          30.000000  162.000000   

       total_phenols  flavanoids  nonflavanoid_phenols  proanthocyanins  \
count     178.000000  178.000000            178.000000       178.000000   
mean        2.295112    2.029270              0.361854         1.590899   
std         0.625

In [43]:
from sklearn.model_selection import train_test_split

# train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(
    features,  # Feature Data
    labels,    # Label Data
    test_size=0.2,  # 테스트 데이터 비율 (20%)
    random_state=42  # 결과 재현을 위한 랜덤 시드
)

print("Train and Test data shapes:")
print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

Train and Test data shapes:
X_train shape: (142, 13)
X_test shape: (36, 13)
y_train shape: (142,)
y_test shape: (36,)


In [44]:
from sklearn.tree import DecisionTreeClassifier

# 모델 생성
decision_tree = DecisionTreeClassifier(random_state=42)

# 학습
decision_tree.fit(X_train, y_train)

# 예측
y_pred = decision_tree.predict(X_test)

# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print(f"Decision Tree Accuracy: {accuracy:.2f}")

Decision Tree Accuracy: 0.94


In [45]:
from sklearn.ensemble import RandomForestClassifier

# 모델 생성
random_forest = RandomForestClassifier(random_state=42)

# 학습
random_forest.fit(X_train, y_train)

# 예측
y_pred = random_forest.predict(X_test)

# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print(f"Random Forest Accuracy: {accuracy:.2f}")

Random Forest Accuracy: 1.00


본 건은 accuracy가 1.0으로 너무 높게 나와 의구심이 들 수 있어 확인을 할 필요가 있다.

In [46]:
from sklearn.svm import SVC

svm = SVC(random_state=42)
svm.fit(X_train, y_train)
y_pred_svm = svm.predict(X_test)

accuracy_svm = accuracy_score(y_test, y_pred_svm)
f1_svm = f1_score(y_test, y_pred_svm, average='macro')
print(f"SVM - Accuracy: {accuracy_svm:.2f}, F1 Score: {f1_svm:.2f}")

SVM - Accuracy: 0.81, F1 Score: 0.76


In [47]:
from sklearn.svm import SVC

svm = SVC(random_state=42)
svm.fit(X_train, y_train)
y_pred_svm = svm.predict(X_test)

accuracy_svm = accuracy_score(y_test, y_pred_svm)
f1_svm = f1_score(y_test, y_pred_svm, average='macro')
print(f"SVM - Accuracy: {accuracy_svm:.2f}, F1 Score: {f1_svm:.2f}")

SVM - Accuracy: 0.81, F1 Score: 0.76


In [48]:
from sklearn.linear_model import SGDClassifier

sgd = SGDClassifier(random_state=42)
sgd.fit(X_train, y_train)
y_pred_sgd = sgd.predict(X_test)

accuracy_sgd = accuracy_score(y_test, y_pred_sgd)
f1_sgd = f1_score(y_test, y_pred_sgd, average='macro')
print(f"SGD Classifier - Accuracy: {accuracy_sgd:.2f}, F1 Score: {f1_sgd:.2f}")

SGD Classifier - Accuracy: 0.72, F1 Score: 0.54


In [49]:
# 결과 저장
results = {
    "Decision Tree": f1_dt,
    "Random Forest": f1_rf,
    "SVM": f1_svm,
    "SGD Classifier": f1_sgd,
    "Logistic Regression": f1_lr
}

best_model = max(results, key=results.get)
print("\nBest Model:")
print(f"{best_model} with F1 Score: {results[best_model]:.2f}")


Best Model:
Logistic Regression with F1 Score: 0.98


와인 data 는 로지스틱 분석방법이 0.98로 높게 나왔기 때문에 로지스틱 분석을 하는 것이 바람직하다.